src/fbas_analyze.rs

```diff
@@ -0,0 +1,253 @@
+// Tseitin transformation on the constrains into "ANDs of ORs", which also
+// introduces additional imaginary atomics. Once the solver produces a
+// satisfiable result (result == SAT), that means a disjoint quorum has been
+// found.
+
+struct FbasLitsWrapper {
+    vertex_count: usize,
+}
+
+impl FbasLitsWrapper {
+    fn new(vcount: usize) -> Self {
+        Self {
+            vertex_count: vcount,
+        }
+    }
+
+    fn in_quorum_a(&self, ni: &NodeIndex) -> Lit {
+        Lit::new(Var::from_index(ni.index()), true)
+    }
+
+    fn in_quorum_b(&self, ni: &NodeIndex) -> Lit {
+        Lit::new(Var::from_index(ni.index() + self.vertex_count), true)
+    }
+
+    fn new_proposition<Solver: SolverInterface>(&self, solver: &mut Solver) -> Lit {
+        Lit::new(solver.new_var_default(), true)
+    }
+}
+
+#[derive(Default)]
+pub struct FbasAnalyzer<Cb: Callbacks> {
+    fbas: Fbas,
+    solver: Solver<Cb>,
+    status: SolveStatus,
+}
+
+#[derive(Clone, Default, PartialEq)]
+pub enum SolveStatus {
+    SAT((Vec<NodeIndex>, Vec<NodeIndex>)),
+    UNSAT,
+    #[default]
+    UNKNOWN,
+}
+
+impl std::fmt::Debug for SolveStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SolveStatus::SAT((quorum_a, quorum_b)) => {
+                write!(f, "SAT(quorum_a: {:?}, quorum_b: {:?})", quorum_a, quorum_b)
+            }
+            SolveStatus::UNSAT => write!(f, "UNSAT"),
+            SolveStatus::UNKNOWN => write!(f, "UNKNOWN"),
+        }
+    }
+}
+
+impl std::fmt::Display for SolveStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        <Self as std::fmt::Debug>::fmt(&self, f)
+    }
+}
+
+impl<Cb: Callbacks> FbasAnalyzer<Cb> {
+    pub fn from_quorum_set_map_buf<T: AsRef<[u8]>, I: ExactSizeIterator<Item = T>>(
+        nodes: I,
+        quorum_set: I,
+        cb: Cb,
+    ) -> Result<Self, FbasError> {
+        let fbas = Fbas::from_quorum_set_map_buf(nodes, quorum_set)?;
+        Self::from_fbas(fbas, cb)
+    }
+
+    #[cfg(any(feature = "json", test))]
+    pub fn from_json_path(path: &str, cb: Cb) -> Result<Self, FbasError> {
+        let fbas = Fbas::from_json_path(path)?;
+        Self::from_fbas(fbas, cb)
+    }
+
+    pub(crate) fn from_fbas(fbas: Fbas, cb: Cb) -> Result<Self, FbasError> {
+        let mut analyzer = Self {
+            fbas,
+            solver: Solver::new(Default::default(), cb),
+            status: SolveStatus::UNKNOWN,
+        };
+        analyzer.construct_formula()?;
+        Ok(analyzer)
+    }
+
+    fn construct_formula(&mut self) -> Result<(), FbasError> {
+        let fbas = &self.fbas;
+        let fbas_lits = FbasLitsWrapper::new(fbas.graph.node_count());
+
+        // for each vertex in the graph, we add a variable representing it
+        // belonging to quorum A and quorum B
+        fbas.graph.node_indices().for_each(|_| {
+            self.solver.new_var_default();
```

Comment:
Given this and the implementations of in_quorum_a, in_quorum_b, it looks like the code is relying on some internal details of batsat, namely that the variable index will start at 0 and will be incremented only by 1 each time a new var is created. Would it not be better to populate a map from graph vertice to batsat Var, in order not to depend on the internal numbering used by batsat?